### ISSUE TITLE
V9.0.0 - Difference in handling of DateTime of 'Unspecified' kind from 8.7.0

### ISSUE THREAD

**Issue by kristianhald** [Azure/azure-storage-net#634]
Upgraded projects from version 8.7.0 to 9.0.0 and saw a difference in handling of DateTimes with 'Unspecified' kind, where in version 8.7.0 'Unspecified' seems to be translated to 'Utc', while in version 9.0.0 it is translated to Local (I am in Denmark with a current of +01 on UTC).

Version | Before Storage DateTime | After Storage DateTime
--- | --- | ---
8.7.0 | 2/15/2018 16:00:00 Unspecified | **2/15/2018 16:00:00 Utc**
9.0.0 | 2/15/2018 16:00:00 Unspecified | **2/15/2018 15:00:00 Utc**
8.7.0 | 2/15/2018 16:00:00 Local | 2/15/2018 15:00:00 Utc
9.0.0 | 2/15/2018 16:00:00 Local | 2/15/2018 15:00:00 Utc
8.7.0 | 2/15/2018 16:00:00 Utc | 2/15/2018 16:00:00 Utc
9.0.0 | 2/15/2018 16:00:00 Utc | 2/15/2018 16:00:00 Utc

The issue is only related to DateTime where the kind is 'Unspecified', as both 'Local' and 'Utc' are handled identical. We are running .NET Core 2.0.

*PS C:\Projects\Mover\Logistics\logistics> dotnet --version*
*2.1.4*

This issue does not affect us, as we must always store in Utc. The only reason we encountered this error, is because we use AutoFixture for testing and it automatically created a DateTime with Kind 'Unspecified' and after upgrading the test began failing.

I have added a quick application, which I used to get above results.

```c#
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.RetryPolicies;
using Microsoft.WindowsAzure.Storage.Table;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using C = System.Console;

namespace Console
{
    class Program
    {
        public class DataEntity : TableEntity
        {
            public DataEntity()
            { }

            public DataEntity(
                Guid id,
                DateTime dateTime,
                DateTimeOffset dateTimeOffset)
                : base(CreatePartitionKey(id), CreateRowKey())
            {
                Id = id;
                DateTime = dateTime;
                DateTimeOffset = dateTimeOffset;
            }

            public Guid Id { get; set; }

            public DateTime DateTime { get; set; }

            public DateTimeOffset DateTimeOffset { get; set; }

            public static string CreatePartitionKey(Guid id)
            {
                return $"{id.ToString("n")}";
            }

            public static string CreateRowKey()
            {
                return $"{DateTime.UtcNow.Ticks}";
            }
        }

        static void Main(string[] args)
        {
            var beforeEntity = new DataEntity(
                Guid.NewGuid(),
                new DateTime(2018, 02, 15, 16, 00, 00, DateTimeKind.Unspecified),
                new DateTimeOffset(new DateTime(2018, 02, 15, 16, 00, 00, DateTimeKind.Unspecified)));
            C.WriteLine($"{"Before".PadRight(10)}{beforeEntity.Id} - {beforeEntity.DateTime} {beforeEntity.DateTime.Kind} - {beforeEntity.DateTimeOffset}");

            var cloudstorage = CloudStorageAccount.Parse("<Your ConnectionString>"); <--- Remember to add your connection string here
            StoreAsync(cloudstorage, beforeEntity).Wait();

            var afterEntity = ReadAsync(cloudstorage, beforeEntity.Id).Result.First();
            C.WriteLine($"{"After".PadRight(10)}{afterEntity.Id} - {afterEntity.DateTime} {afterEntity.DateTime.Kind} - {afterEntity.DateTimeOffset}");

            C.ReadKey();
        }

        public static async Task StoreAsync(CloudStorageAccount cloudstorage, DataEntity entity)
        {
            var tableClient = cloudstorage.CreateCloudTableClient();
            var table = tableClient.GetTableReference("Test");
            await table.CreateIfNotExistsAsync();

            var batchOperation = new TableBatchOperation();
            batchOperation.Insert(entity);
            var results = await table.ExecuteBatchAsync(batchOperation);

            if (results.Any(result => result.HttpStatusCode != 204))
                throw new Exception($"Could not insert entry, because of error while storing.");
        }

        public static async Task<IEnumerable<DataEntity>> ReadAsync(CloudStorageAccount cloudstorage, Guid id)
        {
            CloudTableClient tableClient = cloudstorage.CreateCloudTableClient();
            CloudTable table = tableClient.GetTableReference("Test");
            await table.CreateIfNotExistsAsync();

            TableContinuationToken continuationToken = null;
            var entities = new List<DataEntity>();
            do
            {
                var query = new TableQuery<DataEntity>()
                    .Where(TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.Equal, DataEntity.CreatePartitionKey(id)));

                var options = new TableRequestOptions
                {
                    MaximumExecutionTime = TimeSpan.FromSeconds(15),
                    RetryPolicy = new LinearRetry(TimeSpan.FromMilliseconds(150), 3)
                };

                var result = await table.ExecuteQuerySegmentedAsync(
                    query,
                    continuationToken,
                    options,
                    new OperationContext());

                entities.AddRange(result.Results);
                continuationToken = result.ContinuationToken;
            } while (continuationToken != null);

            return entities;
        }
    }
}
```

**Comment by ajayshiva** [comment]
We are facing the same issue. While saving a local time in azure table it converts it to UTC timezone

**Comment by erezvani1529** [comment]
Thanks @kristianhald for reporting this issue. We will update with more details about a possible fix in v9.2.

@ajayshiva, what you have described seems to be the expected behavior for v8.7 and is not the same issue as what has been reported originally. Did you start seeing this issue after upgrade to 9.0?

**Comment by ajayshiva** [comment]
Now that 9.2 is out, do you have any update on this issue?

We are unable to save local date time using storage sdk in table storage

**Comment by kristianhald** [comment]
Hi @ajayshiva,

I tested using the Console program I added.

When running with version 8.7.0 I get the following results:
Before    3062d2c6-f57c-4f98-bdaf-28acb2d9d920 - 2/15/2018 16:00:00 Unspecified - 2/15/2018 16:00:00 **+01:00**
After     3062d2c6-f57c-4f98-bdaf-28acb2d9d920 - 2/15/2018 **16:00:00 Utc** - 2/15/2018 15:00:00 +00:00

When running the same program with version 9.2.0 I get this result:
Before    92b18859-1f66-44a3-9133-3c7a52c4b92b - 2/15/2018 16:00:00 Unspecified - 2/15/2018 16:00:00 **+01:00**
After     92b18859-1f66-44a3-9133-3c7a52c4b92b - 2/15/2018 **15:00:00 Utc** - 2/15/2018 15:00:00 +00:00

I expect the time being the same across versions, unless you on purpose wanted to make a breaking change between major versions, which is ok if it is documented. :)

With 8.7.0 when reading Unspecified it is converted to Utc but keeps the time as 16:00:00. In 8.7.0 the Unspecified kind is treated as Utc.
However, in version 9.x.x Unspecified is converted to Local and then stored as Utc in the storage, which means the time is different. In 9.x.x the Unspecified kind is treated as Local. 

Now that I look at DateTimeOffset (The second timestamp in the lines) it seems that it has always treated Unspecified as Local (See the +01:00 on the end), so how version 9.x.x. works is probably the correct solution, when looking across the date times; but it is still a breaking change compared to 8.7.0 and might give problems for some systems if they upgrade.

I hope this input helps and thank you for your time :)
